In [1]:
pip install pickleshare

Note: you may need to restart the kernel to use updated packages.


## Question 1

* Install `uv`
* What's the version of uv you installed?
* Use `--version` to find out

In [2]:
pip install uv

Note: you may need to restart the kernel to use updated packages.


In [3]:
!uv --version

uv 0.9.5 (d5f39331a 2025-10-21)


In [4]:
!uv add scikit-learn


Resolved 6 packages in 21ms
Audited 5 packages in 1ms


In [5]:
cd "C:\Users\HP\Desktop\Machine_Learning_Zoomcamp\machine-learning-zoomcamp-master\05-deployment"

C:\Users\HP\Desktop\Machine_Learning_Zoomcamp\machine-learning-zoomcamp-master\05-deployment



## Question 2

* Use uv to install Scikit-Learn version 1.6.1 
* What's the first hash for Scikit-Learn you get in the lock file?
* Include the entire string starting with sha256:, don't include quotes

In [7]:
with open("uv.lock", "r") as f:
    for line in f:
        if "sha256" in line:
            print(line.strip())
            break

sdist = { url = "https://files.pythonhosted.org/packages/e8/5d/447af5ea094b9e4c4054f82e223ada074c552335b9b4b2d14bd9b35a67c4/joblib-1.5.2.tar.gz", hash = "sha256:3faa5c39054b2f03ca547da9b2f52fde67c06240c31853f306aea97f13647b55", size = 331077, upload-time = "2025-08-27T12:15:46.575Z" }


## Question 3

Let's use the model!

* Write a script for loading the pipeline with pickle
* Score this record:

```json
{
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}
```

What's the probability that this lead will convert? 

* 0.333
* 0.533
* 0.733
* 0.933

If you're getting errors when unpickling the files, check their checksum:

```bash
$ md5sum pipeline_v1.bin
7d17d2e4dfbaf1e408e1a62e6e880d49 *pipeline_v1.bin
```


In [10]:
!uv run python predict_one.py

Probability that this lead will convert: 0.5336072702798061


In [11]:
!uv run python predict_two.py

Probability that this lead will convert: 0.5336072702798061


## Question 4

Now let's serve this model as a web service

* Install FastAPI
* Write FastAPI code for serving the model
* Now score this client using `requests`:

```python
url = "YOUR_URL"
client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}
requests.post(url, json=client).json()
```

What's the probability that this client will get a subscription?

* 0.334
* 0.534
* 0.734
* 0.934


In [14]:
!uv add fastapi uvicorn

Resolved 21 packages in 9ms
Audited 20 packages in 17ms


In [13]:
!uv run uvicorn predict_service:app --reload


^C


In [15]:
import requests

url = "http://127.0.0.1:8000/predict"

client = {
    "lead_source": "organic_search",
    "number_of_courses_viewed": 4,
    "annual_income": 80304.0
}

response = requests.post(url, json=client)
print(response.json())

{'subscription_probability': 0.5340417283801275}



## Question 5

Download the base image `agrigorev/zoomcamp-model:2025`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

So what's the size of this base image?

* 45 MB
* 121 MB
* 245 MB
* 330 MB

You can get this information when running `docker images` - it'll be in the "SIZE" column.